[WebGPU] Buffer sizes for pipelines need to be decoupled from the underlying layouts (274098)

mwyrzykowski opened the pull request:
#### f27603377e0a3151445d1d0f986f5f3dee6b05a9
<pre>
[WebGPU] Buffer sizes for pipelines need to be decoupled from the underlying layouts (274098)
<a href="https://bugs.webkit.org/show_bug.cgi?id=274098">https://bugs.webkit.org/show_bug.cgi?id=274098</a>
&lt;radar://128011723&gt;

Reviewed by Tadeu Zagallo.

The same bind group layout can be used in multiple pipelines,
so we should store the information for buffer sizes in the pipeline
and not the layout itself.

* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::executePreDispatchCommands):
* Source/WebGPU/WebGPU/ComputePipeline.h:
(WebGPU::ComputePipeline::create):
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::ComputePipeline::ComputePipeline):
(WebGPU::ComputePipeline::bufferSizes const):
* Source/WebGPU/WebGPU/Pipeline.h:
* Source/WebGPU/WebGPU/Pipeline.mm:
(WebGPU::createLibrary):
(WebGPU::errorValidatingBindGroup):
(WebGPU::validateBindGroup): Deleted.
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::executePreDrawCommands):
* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::RenderPipeline::RenderPipeline):
(WebGPU::m_bufferSizes):
(WebGPU::RenderPipeline::bufferSizes const):

Canonical link: <a href="https://commits.webkit.org/278794@main">https://commits.webkit.org/278794@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03b5067d4a32aa2eed0e9571cc2647319e68cd08

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/51557 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30867 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3909 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54824 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/2250 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/37205 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1930 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41977 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/53656 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/28514 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/44461 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/23117 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25809 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1722 "Passed tests") | [⏳ 🛠 wpe-skia ](https://ews-build.webkit.org/#/builders/WPE-Skia-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/47788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1808 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/56416 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/26677 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/1688 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/49369 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27916 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/44531 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/48560 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11287 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28810 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/27651 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebGPU/WebGPU/ComputePipeline.mm:
```diff
@@ -175,6 +178,12 @@ static MTLSize metalSize(auto workgroupSize, const HashMap<String, WGSL::Constan
+const BufferBindingSizesForBindGroup* ComputePipeline::bufferSizes(uint32_t index) const
+{
+    auto it = m_bufferSizes.find(index);
+    return it == m_bufferSizes.end() ? nullptr : &it->value;
```
just to make sure I understand it: this might be null when the shader doesn't use any bindings in this bind group?

mwyrzykowski replied:
or if there is a mismatch between the number of bind groups in the pipeline and the number used in the render pass. That may fail earlier due to existing validation, but did not want to rely on that

tadeuzagallo commented on Source/WebGPU/WebGPU/Pipeline.h:
```diff
@@ -31,16 +31,19 @@ namespace WebGPU {
 class BindGroup;
 class ShaderModule;
 
+using BufferBindingSizesForBindGroup = HashMap<uint32_t, uint64_t, DefaultHash<uint32_t>, WTF::UnsignedWithZeroKeyHashTraits<uint32_t>>;
+using BufferBindingSizesForPipeline = HashMap<uint32_t, BufferBindingSizesForBindGroup, DefaultHash<uint32_t>, WTF::UnsignedWithZeroKeyHashTraits<uint32_t>>;
```
A general (slightly nit-picky) comment: I think all these types and variables should say "MinBufferBindingSize...", since it's not the actual size of the buffer

tadeuzagallo commented on Source/WebGPU/WebGPU/Pipeline.mm:
```diff
@@ -183,20 +183,25 @@ bool validateBindGroup(BindGroup& bindGroup)
+                return [NSString stringWithFormat:@"Buffer size is missing for binding at index %u bind group", bindingIndex];
+
+            uint64_t bufferSize = 0;
+            if (bufferSizes) {
```
I don't think this logic is correct. I think the user-provided value takes precedence over the computed value, but we're still missing the check that the user-provided value is not smaller than the computed value.

mwyrzykowski replied:
I think that’s right too, I’ll fix that along with the naming

mwyrzykowski committed:
[WebGPU] Buffer sizes for pipelines need to be decoupled from the underlying layouts (274098)
https://bugs.webkit.org/show_bug.cgi?id=274098
<radar://128011723>

Reviewed by Tadeu Zagallo.

The same bind group layout can be used in multiple pipelines,
so we should store the information for buffer sizes in the pipeline
and not the layout itself.

* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::executePreDispatchCommands):
* Source/WebGPU/WebGPU/ComputePipeline.h:
(WebGPU::ComputePipeline::create):
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::ComputePipeline::ComputePipeline):
(WebGPU::ComputePipeline::bufferSizes const):
* Source/WebGPU/WebGPU/Pipeline.h:
* Source/WebGPU/WebGPU/Pipeline.mm:
(WebGPU::createLibrary):
(WebGPU::errorValidatingBindGroup):
(WebGPU::validateBindGroup): Deleted.
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::executePreDrawCommands):
* Source/WebGPU/WebGPU/RenderPipeline.h:
(WebGPU::RenderPipeline::create):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::RenderPipeline::RenderPipeline):
(WebGPU::m_bufferSizes):
(WebGPU::RenderPipeline::bufferSizes const):

Canonical link: https://commits.webkit.org/278794@main
